File: python/Yacht.py
from collections import Counter
YACHT = (lambda x: 50 if len(set(x)) == 1 else 0)
ONES = (lambda x: digits(x,1))
TWOS = (lambda x: digits(x,2))
THREES = (lambda x: digits(x,3))
FOURS = (lambda x: digits(x,4))
FIVES = (lambda x: digits(x,5))
SIXES =(lambda x: digits(x,6))
FULL_HOUSE = (lambda x: sum(x) if sorted(Counter(x).values()) == [2,3] else 0)
FOUR_OF_A_KIND = (lambda x: four_of_a_kind(x)) 
LITTLE_STRAIGHT = (lambda x: 30 if sorted(x) == [1,2,3,4,5] else 0)
BIG_STRAIGHT = (lambda x: 30 if sorted(x) == [2,3,4,5,6] else 0)
CHOICE = sum 

def digits(x,digit):
    return digit*x.count(digit)

def four_of_a_kind(x):
    four_times_elements = [dice for dice in set(x) if x.count(dice) >= 4]
    return 4*four_times_elements[0] if len(four_times_elements) > 0 else 0
    
def score(dice, category):
    return category(dice)

File: javascript/Yacht.js
export const score = (dice, category) => {

  const sides = [1,2,3,4,5,6];
  const count = n => dice.filter(side => side === n).length
  const counts = sides.map(count);
  const sum = dice.reduce((sum, die) => sum += die, 0);
  const product = dice.reduce((product, die) => product *= die, 1);
  const sdice = dice.sort();

  const rules = {
    'ones'            : () => counts[0],
    'twos'            : () => counts[1] * 2,
    'threes'          : () => counts[2] * 3,
    'fours'           : () => counts[3] * 4,
    'fives'           : () => counts[4] * 5,
    'sixes'           : () => counts[5] * 6,
    'full house'      : () => [2,3].every(n => counts.includes(n)) ? sum : 0,
    'four of a kind'  : () => ((counts.indexOf(4)+1) || (counts.indexOf(5)+1)) * 4,
    'little straight' : () => product === 120 ? 30 : 0,
    'big straight'    : () => product === 720 ? 30 : 0,
    'choice'          : () => sum,
    'yacht'           : () => sdice[0] === sdice[4] ? 50 : 0
  };

  return rules[category]();
};